File: pipeline/combine.py
"""Combine income + broadband and compute derived policy metrics."""
from __future__ import annotations

import geopandas as gpd
import pandas as pd

from pipeline import config


def _tercile(series: pd.Series) -> pd.Series:
    """Return 1/2/3 tercile labels by rank (robust to ties and NaN)."""
    ranks = series.rank(method="first")
    return pd.qcut(ranks, 3, labels=[1, 2, 3]).astype("Int64")


def add_derived_metrics(df: pd.DataFrame) -> pd.DataFrame:
    """Add income_speed_ratio and bivariate_class (income-speed terciles)."""
    out = df.copy()
    out["income_speed_ratio"] = out["mean_income"] / out["download_mbps"].where(
        out["download_mbps"] > 0
    )
    inc = _tercile(out["mean_income"])
    spd = _tercile(out["download_mbps"])
    out["bivariate_class"] = inc.astype(str) + "-" + spd.astype(str)
    return out


def main() -> None:
    civ = gpd.read_file(config.CIVIC_ASSOC)[["geoid", "region_name", "geometry"]]
    income = pd.read_csv(config.CIVIC_INCOME, dtype={"geoid": str})
    broadband = pd.read_csv(config.CIVIC_BROADBAND, dtype={"geoid": str})
    merged = civ.merge(income, on="geoid", how="left").merge(
        broadband, on="geoid", how="left"
    )
    merged = add_derived_metrics(merged)
    gdf = gpd.GeoDataFrame(merged, geometry="geometry", crs=civ.crs)
    gdf.to_file(config.CIVIC_COMBINED, driver="GeoJSON")
    print(f"Wrote {len(gdf)} combined civic-association rows → {config.CIVIC_COMBINED}")


if __name__ == "__main__":
    main()

File: pipeline-r/combine.R
# Join redistributed income + broadband onto civic polygons; derive metrics.
# Parallels pipeline/combine.py.
source("pipeline-r/config.R")
suppressPackageStartupMessages(library(sf))

tercile <- function(x) {
  # 1/2/3 labels by rank (robust to ties and NA), mirroring pandas qcut on ranks.
  r <- rank(x, ties.method = "first", na.last = "keep")
  br <- stats::quantile(r, probs = c(0, 1/3, 2/3, 1), na.rm = TRUE)
  cut(r, breaks = br, labels = c("1", "2", "3"), include.lowest = TRUE)
}

add_derived_metrics <- function(df) {
  df$income_speed_ratio <- ifelse(df$download_mbps > 0,
                                  df$mean_income / df$download_mbps, NA_real_)
  inc <- tercile(df$mean_income)
  spd <- tercile(df$download_mbps)
  df$bivariate_class <- paste0(as.character(inc), "-", as.character(spd))
  df
}

run_combine <- function() {
  civ <- sf::st_read(CIVIC_ASSOC, quiet = TRUE)[, c("geoid", "region_name")]
  income    <- read.csv(CIVIC_INCOME_R,    colClasses = c(geoid = "character"))
  broadband <- read.csv(CIVIC_BROADBAND_R, colClasses = c(geoid = "character"))
  civ$geoid <- as.character(civ$geoid)

  merged <- merge(civ, income, by = "geoid", all.x = TRUE)
  merged <- merge(merged, broadband, by = "geoid", all.x = TRUE)
  merged <- add_derived_metrics(merged)

  sf::st_write(merged, CIVIC_COMBINED_R, delete_dsn = TRUE, quiet = TRUE)
  cat(sprintf("Wrote %d combined civic-association rows -> %s\n",
              nrow(merged), CIVIC_COMBINED_R))
}

if (sys.nframe() == 0) run_combine()
